kernel/src/schema/schema_compare.rs

```diff
@@ -0,0 +1,398 @@
+//! Provides utilities to perform comparisons between a [`Schema`]s. The api used to check schema
+//! compatibility is [`can_read_as`].
+//!
+//! # Examples
+//!  ```rust, ignore
+//!  # use delta_kernel::schema::StructType;
+//!  # use delta_kernel::schema::StructField;
+//!  # use delta_kernel::schema::DataType;
+//!  let schema = StructType::new([
+//!     StructField::new("id", DataType::LONG, false),
+//!     StructField::new("value", DataType::STRING, true),
+//!  ]);
+//!  let read_schema = StructType::new([
+//!     StructField::new("id", DataType::LONG, true),
+//!     StructField::new("value", DataType::STRING, true),
+//!     StructField::new("year", DataType::INTEGER, true),
+//!  ]);
+//!  // Schemas are compatible since the `year` value is nullable
+//!  assert!(schema.can_read_as(&read_schema).is_ok());
+//!  ````
+//!
+//! [`Schema`]: crate::schema::Schema
+use std::collections::{HashMap, HashSet};
+
+use crate::utils::require;
+
+use super::{DataType, StructField, StructType};
+
+/// Represents the ways a schema comparison can fail.
+#[derive(Debug, thiserror::Error)]
+pub(crate) enum Error {
+    #[error("The nullability was tightened for a schema field")]
+    NullabilityTightening,
+    #[error("Field names do not match")]
+    FieldNameMismatch,
+    #[error("Schema is invalid")]
+    InvalidSchema,
+    #[error("The read schema is missing a column present in the schema")]
+    MissingColumn,
+    #[error("Read schema has a non-nullable column that is not present in the schema")]
+    NewNonNullableColumn,
+    #[error("Types for two schema fields did not match")]
+    TypeMismatch,
+}
+
+/// A [`std::result::Result`] that has the schema comparison [`Error`] as the error variant.
+#[allow(unused)]
+pub(crate) type SchemaComparisonResult = Result<(), Error>;
+
+/// Represents a schema compatibility check for the type. If `self` can be read as `read_type`,
+/// this function returns `Ok(())`. Otherwise, this function returns `Err`.
+#[allow(unused)]
+pub(crate) trait SchemaComparison {
+    fn can_read_as(&self, read_type: &Self) -> SchemaComparisonResult;
+}
+
+/// The nullability flag of a schema's field. This can be compared with a read schema field's
+/// nullability flag using [`NullabilityFlag::can_read_as`].
+#[allow(unused)]
+pub(crate) struct NullabilityFlag(bool);
+
+impl SchemaComparison for NullabilityFlag {
+    /// Represents a nullability comparison between two schemas' fields. Returns true if the
+    /// read nullability is the same or wider than the nullability of self.
+    fn can_read_as(&self, read_nullable: &NullabilityFlag) -> SchemaComparisonResult {
```

Comment:
Would this be helpful?
```suggestion
    fn can_read_as(&self, read_nullable: impl Into<NullabilityFlag>) -> SchemaComparisonResult {
```
Simplifies callers a bit without sacrificing clarity, e.g.
```rust
NullabilityFlag(self.nullable).can_read_as(read_field.nullable)?;
```

Reply:
See [this](https://github.com/delta-io/delta-kernel-rs/pull/554#discussion_r1924519225)